src/backend/vulkan/src/window.rs

```diff
@@ -34,69 +38,124 @@ impl Drop for RawSurface {
+                s_type: vk::StructureType::XlibSurfaceCreateInfoKhr,
+                p_next: ptr::null(),
+                flags: vk::XlibSurfaceCreateFlagsKHR::empty(),
+                window: unsafe { transmute(window) },
+                dpy: display as *mut _,
+            };
+
+            unsafe { xlib_loader.create_xlib_surface_khr(&info, None) }
+                .expect("XlibSurface failed")
+        };
+
+        let (width, height) = unsafe {
+            use x11::xlib::XGetGeometry;
+            use std::mem::{uninitialized, transmute};
+            use std::os::raw::{c_int, c_uint, c_ulong};
+            let mut root_return: c_ulong = uninitialized();
+            let mut x_return: c_int = uninitialized();
+            let mut y_return: c_int = uninitialized();
+            let mut w_return: c_uint = uninitialized();
+            let mut h_return: c_uint = uninitialized();
+            let mut b_return: c_uint = uninitialized();
+            let mut d_return: c_uint = uninitialized();
+            let result = XGetGeometry(display as *mut _, transmute(window),
+                &mut root_return,
+                &mut x_return, &mut y_return,
+                &mut w_return, &mut h_return,
+                &mut b_return, &mut d_return);
+            if result == 0 {
+                panic!("XGetGeometry failed");
+            }
+            (w_return as u32, h_return as u32)
+        };
+
+        self.create_surface_from_vk_surface_khr(surface, width, height)
+    }
+
+    #[cfg(unix)]
+    pub fn create_surface_from_wayland(&self, display: *mut c_void, surface: *mut c_void, shell_surface: *mut c_void) -> Surface {
+        unimplemented!()
+    }
+
+    #[cfg(target_os = "linux")]
+    pub fn create_surface_from_drm_kms(&self) -> Surface {
+        unimplemented!()
+    }
+
+    #[cfg(windows)]
+    pub fn create_surface_from_hwnd(&self, hwnd: *mut c_void) -> Surface {
+        let entry = VK_ENTRY
+            .as_ref()
+            .expect("Unable to load Vulkan entry points");
+
+        let surface = {
+            use kernel32;
+
+            let win32_loader = ext::Win32Surface::new(entry, &self.raw.0)
+                .expect("Unable to load win32 surface functions");
+
+            unsafe {
+                let info = vk::Win32SurfaceCreateInfoKHR {
+                    s_type: vk::StructureType::Win32SurfaceCreateInfoKhr,
+                    p_next: ptr::null(),
+                    flags: vk::Win32SurfaceCreateFlagsKHR::empty(),
+                    hinstance: kernel32::GetModuleHandleW(ptr::null()) as *mut _,
+                    hwnd: hwnd as *mut _,
+                };
+
+                win32_loader.create_win32_surface_khr(&info, None).ok()
+            }
+        };
+
+        let (width, height) = unsafe {
+            use winapi::RECT;
+            use user32::GetClientRect;
+            use std::mem::uninitialized;
+            let mut rect: RECT = uninitialized();
```

Comment:
why not `mem::zeroed`?

Reply:
IMO `rect` must be filled by `GetClientRect`, and therefore we should leave it undefined on initialization.

Reply:
I don't agree. I don't see a reason why `uninitialized` would be any better then `zeroed` here.

Reply:
As above, I misthought it and corrected the code with `zeroed()`